test/markup/cpp/templates.txt

```diff
@@ -0,0 +1,104 @@
+template <template <typename X> class T> void f();
+
+template class A::B<1, 2> {};
+
+template void A::f(int);
+
+template <class T, std::size_t N = sizeof(T), class K = T>
+ostream& operator<< (ostream& os, const skg::Triplet<T> pt) ;
+void other_stuff_that_isnt_colored();
+
+
+template<class T>
+class My_vector { /* ... */ };
+
+template<class T = void>
+struct My_op_functor { /* ... */ };
+
+template<typename... Ts>
+class My_tuple { /* ... */ };
+
+
+template<class> class My_vector;
+template<class = void> struct My_op_functor;
+template<typename...> class My_tuple;
+
+
+template<typename T> concept C1 = true;
+template<typename... Ts> concept C2 = true; // variadic concept
+template<typename T, typename U> concept C3 = true;
+ 
+template<C1 T> struct s1;         // constraint-expression is C1<T>
+template<C1... T> struct s2;      // constraint-expression is (C1<T> && ...)
+template<C2... T> struct s3;      // constraint-expression is (C2<T> && ...)
+template<C3<int> T> struct s4;    // constraint-expression is C3<T, int>
+template<C3<int>... T> struct s5; // constraint-expression is (C3<T, int> && ...)
+
+
+template<typename K, typename V, template<typename> typename C = my_array>
+class Map {};
+
+
+template<class B>
+template<class C>
+void A<B>::g(C) {}
+
+
+void g() {
+    f<int()>();
+}
+
+
+template<int (&pa)[5]> struct W {};
+template<void (*pf)(int)> struct A {};
+
+
+template<class T, const char* p> class X {};
+
+
+template<typename T> class A { int x; }; // primary template
+template<typename T> class A<T*> { long x; }; // partial specialization
+
+
+template<typename T> struct eval; // primary template 
+ 
+template<template<typename, typename...> class TT, typename T1, typename... Rest>
+struct eval<TT<T1, Rest...>> {}; // partial specialization of eval
+
+
+eval<A<int>> eA;
+eval<B<int, float>> eB;
+eval<C<17>> eC;
+eval<D<int, 17>> eD;
+eval<E<int, float>> eE;
+
+
+template <template <auto> class> void FA();  // note: C++17
+
+
+template<template<typename TT = char> class T>
+void A<T>::g()
+{
+    T<> t; // ok: t is T<char>
+}
+
+s.template foo<T>();
+class C;
+
+
+// Hard mode
+
+template <int a, int b, bool c = (a > b), class T> void f1();
+template <int a, int b, bool c = (a >> b), class T> void f2();
+
+template <int a, int b, bool c = (a < b), class T> void g1();
```

Comment:
This test is failing, and I don’t know why: I thought my rules would handle the nested parenthesised expression — and this works for the corresponding test case above, i.e. for `(a > b)`.

Reply:
I'd guess the `<` in `a < b` is starting a new TEMPLATE, messing everything up.

Reply:
Ah, of course! I actually don’t know how to handle this: I thought about adding `)` to the `end` of the `TEMPLATE_USE`, mode; this works, but then the mode started by `(` won’t terminate, even when I also set `excludeEnd: true`.

Reply:
Adding the following to the `TEMPLATE_USE` mode seems to work:

```javascript
    'on:begin': (matchData, response) => {
      const rest = matchData.input.slice(matchData.index + 1);
      let parens = 0;
      let brackets = 1; // start with `<`
      for (let i = 0; i < rest.length; i++) {
        const c = rest.charAt(i);
        switch (c) {
          case '(': parens++; break;
          case ')': parens--; break;
          case '<': brackets++; break;
          case '>': brackets--; break;
        }
        if (brackets == 0) return;
        if (parens == -1) {
          response.ignoreMatch();
          return;
        }
      }
    },
```

However, this feels like overkill (potentially also inefficient?); it also *only* works in cases where the comparison expression is nested inside parentheses, i.e. it works for this test case but not the next one.

If this approach is taken at all, we’d probably want to add some emergency stop characters (`;`, `{`) to the `switch`.

Reply:
Yeah, this is too much... and definitely has performance implications.

Reply:
I think the best maybe we can do here is liberal use of endsWithParent to make sure we escape if we hit a `{`}, line end, etc... and try to return to normal matching...